File: src/lib/agents/browserRecipe.ts
/* browserRecipe.ts — Mission D: generic web-surface publishing via a
 * driven browser, with ZERO site-specific knowledge in code.
 *
 * A `BrowserRecipe` is DATA describing how to drive one site: which steps
 * to run (navigate/waitFor/click/fill/upload/assert), which step is the
 * irreversible one, and which guard conditions count as an unexpected
 * screen. This module only knows the generic step/guard SHAPE — never a
 * specific site's selectors, URLs, or flow. The same runner drives
 * Instagram, another network, or a page with a different content format
 * without any code change: only the recipe data differs.
 *
 * Talks to the Rust commands in src-tauri/src/commands/browser_recipe.rs:
 *   - browser_recipe_open(sessionId, profileName, headless)
 *   - browser_recipe_step(sessionId, stepJson, guardsJson)
 *   - browser_recipe_close(sessionId)
 *
 * Safety model (Mission D spec):
 *   - `validateOnly` (default true) is the default VALIDATION mode: the
 *     runner stops BEFORE the step flagged `irreversible`, leaving the
 *     last completed step's screenshot as proof of the state reached.
 *   - A tripped guard or a failed step aborts immediately — never retried,
 *     never looped.
 *   - Credentials are referenced by ENVIRONMENT VARIABLE NAME only
 *     (`secretEnvVar`), resolved on the Rust side; this module never reads,
 *     holds, or logs a literal secret value.
 */

import { invoke } from '@tauri-apps/api/core';
import { errorMessage } from '../errorMessage.js';

// ── Types ──────────────────────────────────────────────────────────

export type BrowserStepKind = 'navigate' | 'waitFor' | 'click' | 'fill' | 'upload' | 'assert';

export type BrowserAssertExpectation = 'visible' | 'hidden' | 'textContains';

/** One step of a recipe. Field applicability depends on `kind` — see
 *  validateBrowserRecipe for exactly which fields each kind requires.
 *  `irreversible: true` marks the step that performs an unrecoverable
 *  action (e.g. the final "Publish" click) — the ONLY step the recipe
 *  author needs to flag; everything else about safety follows from that
 *  single bit. */
export interface BrowserStep {
  id: string;
  kind: BrowserStepKind;
  timeoutMs?: number;
  irreversible?: boolean;
  url?: string;
  selector?: string;
  text?: string;
  value?: string;
  /** Name of an environment variable to resolve as this step's `value` —
   *  mutually exclusive with a literal `value`. Never a literal secret. */
  secretEnvVar?: string;
  filePath?: string;
  expect?: BrowserAssertExpectation;
}

/** A circuit-breaker condition: if it matches the page after ANY step, the
 *  recipe run aborts immediately with `circuitBreakerTripped` set — this is
 *  what makes an unexpected screen (captcha, identity check, changed
 *  layout) a stop signal instead of silently plowing ahead or retrying.
 *  What counts as "unexpected" is entirely up to the recipe author; this
 *  module has no built-in notion of it. */
export interface BrowserGuard {
  label: string;
  selector?: string;
  textContains?: string;
}

export interface BrowserRecipe {
  /** Identifies BOTH the persistent login profile AND the session — the
   *  same profileName always resumes the same logged-in browser profile. */
  profileName: string;
  steps: BrowserStep[];
  guards?: BrowserGuard[];
  headless?: boolean;
}

export interface BrowserStepOutcome {
  id: string;
  kind: BrowserStepKind;
  ok: boolean;
  detail?: string;
  error?: string;
  guardTripped?: string;
  screenshotBase64?: string;
}

export interface BrowserRecipeResult {
  /** True when execution halted just before an `irreversible` step because
   *  `validateOnly` was in effect — the intended default outcome for a
   *  first run of any recipe. */
  stoppedBeforeFinal: boolean;
  /** Set when a guard matched after some step — the honest "an unexpected
   *  screen appeared" signal, never silently swallowed. */
  circuitBreakerTripped: { label: string; afterStepId: string } | null;
  /** Set when the recipe/session could not even be opened, or a step
   *  failed outright (not a guard trip). Absent on a clean run. */
  failure: string | null;
  steps: BrowserStepOutcome[];
}

type ValidationResult = { ok: true } | { ok: false; reason: string };

// ── Recipe validation (boundary check — recipes are external data) ──

const REQUIRED_FIELDS_BY_KIND: Record<BrowserStepKind, string[]> = {
  navigate: ['url'],
  waitFor: [],
  click: [],
  fill: [],
  upload: ['filePath'],
  assert: ['expect'],
};

function validateStep(step: BrowserStep, index: number): string | null {
  if (typeof step.id !== 'string' || step.id.length === 0) {
    return `step[${index}]: missing or non-string "id"`;
  }
  const required = REQUIRED_FIELDS_BY_KIND[step.kind];
  if (required === undefined) {
    return `step[${index}] ("${step.id}"): unknown kind "${String(step.kind)}"`;
  }
  for (const field of required) {
    if (!(field in step) || (step as unknown as Record<string, unknown>)[field] === undefined) {
      return `step[${index}] ("${step.id}"): kind "${step.kind}" requires "${field}"`;
    }
  }
  if ((step.kind === 'waitFor' || step.kind === 'click') && !step.selector && !step.text) {
    return `step[${index}] ("${step.id}"): kind "${step.kind}" requires "selector" or "text"`;
  }
  if (step.kind === 'fill' && !step.selector) {
    return `step[${index}] ("${step.id}"): kind "fill" requires "selector"`;
  }
  if (step.kind === 'fill' && step.value !== undefined && step.secretEnvVar !== undefined) {
    return `step[${index}] ("${step.id}"): "value" and "secretEnvVar" are mutually exclusive`;
  }
  if (step.kind === 'assert' && step.expect === 'textContains' && !step.text) {
    return `step[${index}] ("${step.id}"): expect "textContains" requires "text"`;
  }
  if (step.kind === 'assert' && (step.expect === 'visible' || step.expect === 'hidden') && !step.selector) {
    return `step[${index}] ("${step.id}"): expect "${step.expect}" requires "selector"`;
  }
  return null;
}

/** Validates a recipe before any browser interaction is attempted — recipes
 *  are external data (from a stored artifact or the manager's own
 *  generation), never trusted verbatim. Returns the first problem found. */
export function validateBrowserRecipe(recipe: BrowserRecipe): ValidationResult {
  if (!recipe.profileName || typeof recipe.profileName !== 'string') {
    return { ok: false, reason: 'recipe is missing a non-empty "profileName"' };
  }
  if (!Array.isArray(recipe.steps) || recipe.steps.length === 0) {
    return { ok: false, reason: 'recipe must have at least one step' };
  }
  for (let i = 0; i < recipe.steps.length; i += 1) {
    const problem = validateStep(recipe.steps[i], i);
    if (problem) return { ok: false, reason: problem };
  }
  const guards = recipe.guards ?? [];
  for (let i = 0; i < guards.length; i += 1) {
    const g = guards[i];
    if (!g.label || (!g.selector && !g.textContains)) {
      return { ok: false, reason: `guard[${i}]: requires "label" and one of "selector"/"textContains"` };
    }
  }
  return { ok: true };
}

// ── Step JSON shaping ────────────────────────────────────────────────

/** Strips the recipe-authoring `id` field out of the step payload sent to
 *  Rust — the id is a caller-side bookkeeping concern (matching outcomes
 *  back to steps), not something the controller needs to execute a step. */
function stepPayload(step: BrowserStep): Record<string, unknown> {
  const { id: _id, irreversible: _irreversible, ...rest } = step;
  return rest;
}

interface RawStepResult {
  ok: boolean;
  detail?: string | null;
  error?: string | null;
  guardTripped?: string | null;
  screenshotBase64?: string | null;
}

function parseStepResult(raw: string): RawStepResult {
  try {
    const parsed = JSON.parse(raw) as RawStepResult;
    return {
      ok: parsed.ok === true,
      detail: parsed.detail ?? undefined,
      error: parsed.error ?? undefined,
      guardTripped: parsed.guardTripped ?? undefined,
      screenshotBase64: parsed.screenshotBase64 ?? undefined,
    };
  } catch {
    return { ok: false, error: `Malformed step result: ${raw}` };
  }
}

// ── Execution ────────────────────────────────────────────────────────

export interface RunBrowserRecipeOptions {
  /** Default true: the safe, default validation mode — stop BEFORE the
   *  step flagged `irreversible`, proving the state reached with the last
   *  screenshot instead of performing the irreversible action. */
  validateOnly?: boolean;
}

/** Runs a recipe against a persistent browser profile session. Never
 *  retries a failed step or a tripped guard — either condition aborts the
 *  run immediately and is reported honestly in the result, not swallowed.
 *  Always closes the session on the way out, success or failure. */
export async function runBrowserRecipe(
  recipe: BrowserRecipe,
  opts: RunBrowserRecipeOptions = {},
): Promise<BrowserRecipeResult> {
  const validation = validateBrowserRecipe(recipe);
  if (!validation.ok) {
    return { stoppedBeforeFinal: false, circuitBreakerTripped: null, failure: validation.reason, steps: [] };
  }

  const validateOnly = opts.validateOnly ?? true;
  const sessionId = recipe.profileName;
  const guardsJson = JSON.stringify(recipe.guards ?? []);
  const steps: BrowserStepOutcome[] = [];
  let stoppedBeforeFinal = false;
  let circuitBreakerTripped: BrowserRecipeResult['circuitBreakerTripped'] = null;
  let failure: string | null = null;

  try {
    await invoke<string>('browser_recipe_open', {
      sessionId,
      profileName: recipe.profileName,
      headless: recipe.headless ?? true,
    });
  } catch (err: unknown) {
    // Rust inserts the controller process into its session map BEFORE
    // reporting an "open" failure, so a failed open can still leave a
    // spawned-but-unopened process behind — best-effort close it now
    // rather than waiting for the next same-profile open or app exit.
    try {
      await invoke<string>('browser_recipe_close', { sessionId });
    } catch {
      /* best-effort cleanup only */
    }
    return {
      stoppedBeforeFinal: false,
      circuitBreakerTripped: null,
      failure: `Failed to open recipe session: ${errorMessage(err)}`,
      steps: [],
    };
  }

  try {
    for (const step of recipe.steps) {
      if (step.irreversible && validateOnly) {
        stoppedBeforeFinal = true;
        break;
      }

      let raw: string;
      try {
        raw = await invoke<string>('browser_recipe_step', {
          sessionId,
          stepJson: JSON.stringify(stepPayload(step)),
          guardsJson,
        });
      } catch (err: unknown) {
        steps.push({ id: step.id, kind: step.kind, ok: false, error: errorMessage(err) });
        failure = `Step "${step.id}" failed: ${errorMessage(err)}`;
        break;
      }

      const result = parseStepResult(raw);
      steps.push({
        id: step.id,
        kind: step.kind,
        ok: result.ok,
        detail: result.detail ?? undefined,
        error: result.error ?? undefined,
        guardTripped: result.guardTripped ?? undefined,
        screenshotBase64: result.screenshotBase64 ?? undefined,
      });

      // Circuit breaker or step failure: abort immediately, never retry,
      // never loop — this is the honest stop signal, not a silent skip.
      if (result.guardTripped) {
        circuitBreakerTripped = { label: result.guardTripped, afterStepId: step.id };
        break;
      }
      if (!result.ok) {
        failure = result.error ?? `Step "${step.id}" failed`;
        break;
      }
    }
  } finally {
    // Always close, success or failure — closing itself is best-effort
    // (never overrides the real result computed above with its own error).
    try {
      await invoke<string>('browser_recipe_close', { sessionId });
    } catch {
      /* best-effort cleanup only */
    }
  }

  return { stoppedBeforeFinal, circuitBreakerTripped, failure, steps };
}

File: src-tauri/src/commands/browser_recipe.rs
//! Recipe execution Tauri commands — Mission D, generic web publishing.
//!
//! Drives the SAME Playwright controller process and `instances()` map as
//! `commands::browser` (see that module for the controller script itself),
//! keyed by a caller-chosen `session_id` instead of the hardcoded
//! `BROWSER_INSTANCE_ID`, so a recipe session never collides with a manual
//! browsing session.
//!
//! All site-specific knowledge — which selector to click, which URL to
//! open, what counts as an unexpected screen — is supplied by the CALLER
//! as data: a "recipe" of steps and guards (see
//! `src/lib/agents/browserRecipe.ts`). This module only knows the generic
//! step kinds (`navigate`/`waitFor`/`click`/`fill`/`upload`/`assert`) and
//! generic guard shapes (`selector`/`textContains`) — never a specific
//! site's shape.
//!
//! Commands:
//! - `browser_recipe_open` — launches (or reuses) a persistent-profile
//!   session so a login survives across executions
//! - `browser_recipe_step` — executes exactly one recipe step, time-bounded,
//!   with a guard check + screenshot proof after every step
//! - `browser_recipe_close` — closes the session and kills the controller
//!
//! "Stop before the final action" (the default validation mode) and the
//! circuit-breaker's "never retry" rule are enforced by the TypeScript
//! caller, which decides per step whether to send it at all — this module
//! only executes whatever single step it's asked to run.

use std::process::{Command, Stdio};

use super::browser::{instances, send_and_receive, write_controller_script, BrowserState};

/// Default per-step bound used when a recipe step omits `timeoutMs` — kept
/// in sync with the controller script's own `10000` fallback so the outer
/// transport timeout (`step_timeout_ms + OUTER_TIMEOUT_MARGIN_MS`) is never
/// tighter than the inner one the controller actually enforces.
const DEFAULT_STEP_TIMEOUT_MS: u64 = 10_000;

/// Margin added on top of a step's own timeout for the OUTER transport
/// read (`send_and_receive`) — the controller enforces the inner bound
/// itself (`withTimeout`, +2000ms there) and always replies with a result
/// object either way; this margin only guards against transport overhead.
const OUTER_TIMEOUT_MARGIN_MS: u64 = 5_000;

/// Sanitizes a caller-supplied profile name into a filesystem-safe slug.
/// Recipes are DATA supplied by the manager/user, never trusted verbatim
/// for a path join — an unsanitized name could contain `..`, path
/// separators, or reserved Windows device names. Non `[a-zA-Z0-9_-]`
/// characters are dropped and the result capped at 64 bytes; an
/// all-invalid or empty input falls back to `"default"`.
fn sanitize_profile_name(raw: &str) -> String {
    let cleaned: String = raw
        .chars()
        .filter(|c| c.is_ascii_alphanumeric() || *c == '-' || *c == '_')
        .take(64)
        .collect();
    if cleaned.is_empty() {
        "default".to_string()
    } else {
        cleaned
    }
}

/// Root directory under which persistent browser profiles are stored, one
/// subdirectory per sanitized profile name — this is what gives Mission
/// D's "profil connecté persistant" its persistence across app restarts.
/// Prefers `LAZY_APP_CONFIG_DIR` (the same env var
/// `brain::config::expose_app_config_dir` sets at startup, so profiles
/// live alongside the rest of the app's own data); falls back to the OS
/// temp dir when unset (e.g. this module's own unit tests).
fn browser_profiles_root() -> std::path::PathBuf {
    let base = std::env::var("LAZY_APP_CONFIG_DIR")
        .ok()
        .filter(|v| !v.is_empty())
        .map(std::path::PathBuf::from)
        .unwrap_or_else(std::env::temp_dir);
    base.join("browser-profiles")
}

/// Resolves the persistent profile directory for a given (unsanitized)
/// profile name. Does not create the directory — callers create it lazily
/// right before launching the persistent context.
fn profile_dir(profile_name: &str) -> std::path::PathBuf {
    browser_profiles_root().join(sanitize_profile_name(profile_name))
}

/// Resolves a recipe step's declared `secretEnvVar` (if any) into a
/// literal `value` field, IN PLACE, removing the `secretEnvVar` key
/// afterward. Recipes may only reference a credential by the NAME of an
/// environment variable — never a literal secret — per Mission D's
/// non-negotiable "identifiants passés par variable d'environnement,
/// jamais écrits dans le dépôt". The resolved value is written only into
/// the JSON message sent to the controller's stdin (never logged —
/// `send_and_receive` logs nothing about message content) and is never
/// echoed back in a step result. Fails fast (rather than filling a blank
/// or wrong value) when the named variable isn't set; the error message
/// names the variable, never a value.
fn resolve_step_secret(step: &mut serde_json::Value) -> Result<(), String> {
    let env_var = match step.get("secretEnvVar").and_then(|v| v.as_str()) {
        Some(name) => name.to_string(),
        None => return Ok(()),
    };
    let value = std::env::var(&env_var)
        .map_err(|_| format!("Environment variable '{}' is not set", env_var))?;
    if let Some(obj) = step.as_object_mut() {
        obj.insert("value".to_string(), serde_json::Value::String(value));
        obj.remove("secretEnvVar");
    }
    Ok(())
}

/// Opens a persistent-profile Playwright session for recipe execution,
/// keyed by `session_id` (distinct from the manual `browser_playwright_*`
/// commands' hardcoded `"default"` instance, so the two never collide in
/// the shared `instances()` map). `profile_name` selects which on-disk
/// profile directory to reuse across runs — the SAME profile name always
/// resumes the SAME logged-in session.
#[tauri::command]
pub fn browser_recipe_open(session_id: String, profile_name: String, headless: bool) -> Result<String, String> {
    let dir = profile_dir(&profile_name);
    std::fs::create_dir_all(&dir).map_err(|e| format!("Failed to create profile dir: {}", e))?;

    let mut map = instances().lock().map_err(|e| format!("Lock error: {}", e))?;
    if let Some(mut old) = map.remove(&session_id) {
        let _ = old.child.kill();
        let _ = old.child.wait();
    }

    let script_path = write_controller_script()?;
    let child = Command::new("node")
        .arg(&script_path)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .map_err(|e| format!("Failed to spawn Playwright controller: {}", e))?;
    map.insert(session_id.clone(), BrowserState { child });

    let state = map
        .get_mut(&session_id)
        .ok_or_else(|| "Failed to get recipe session state".to_string())?;

    let msg = serde_json::json!({
        "id": 1,
        "action": "openPersistent",
        "params": { "userDataDir": dir.to_string_lossy(), "headless": headless }
    });
    let response = send_and_receive(state, &msg.to_string(), 30000)?;
    let parsed: serde_json::Value = serde_json::from_str(&response)
        .map_err(|e| format!("Failed to parse controller response: {}", e))?;

    if let Some(err) = parsed.get("error") {
        return Err(format!("Recipe browser open failed: {}", err));
    }

    Ok("Recipe browser session opened".to_string())
}

/// Executes exactly one recipe step against an already-open recipe
/// session. `step_json` / `guards_json` are the recipe's own step and
/// guard definitions, serialized by the TypeScript caller — this command
/// has no built-in notion of what any given step or guard means for a
/// particular site.
///
/// Returns the controller's structured result verbatim as a JSON string:
/// `{ ok, detail, error, guardTripped, screenshotBase64 }`. Never retries
/// — a single failed step or tripped guard is the caller's signal to stop.
#[tauri::command]
pub fn browser_recipe_step(session_id: String, step_json: String, guards_json: String) -> Result<String, String> {
    let mut step: serde_json::Value =
        serde_json::from_str(&step_json).map_err(|e| format!("Invalid step JSON: {}", e))?;
    let guards: serde_json::Value =
        serde_json::from_str(&guards_json).map_err(|e| format!("Invalid guards JSON: {}", e))?;

    resolve_step_secret(&mut step)?;

    let step_timeout_ms = step
        .get("timeoutMs")
        .and_then(|v| v.as_u64())
        .filter(|v| *v > 0)
        .unwrap_or(DEFAULT_STEP_TIMEOUT_MS);
    let outer_timeout_ms = step_timeout_ms + OUTER_TIMEOUT_MARGIN_MS;

    let mut map = instances().lock().map_err(|e| format!("Lock error: {}", e))?;
    let state = map
        .get_mut(&session_id)
        .ok_or_else(|| "Recipe session not open. Call browser_recipe_open first.".to_string())?;

    let msg = serde_json::json!({
        "id": 2,
        "action": "runStep",
        "params": { "step": step, "guards": guards }
    });
    let response = send_and_receive(state, &msg.to_string(), outer_timeout_ms)?;
    let parsed: serde_json::Value = serde_json::from_str(&response)
        .map_err(|e| format!("Failed to parse response: {}", e))?;

    if let Some(err) = parsed.get("error") {
        return Err(format!("Recipe step transport failed: {}", err));
    }

    let result = parsed.get("result").cloned().unwrap_or(serde_json::Value::Null);
    Ok(result.to_string())
}

/// Closes a recipe session and kills its controller process. Same
/// best-effort-graceful-then-tree-kill-on-drop shape as
/// `browser_playwright_close` in `commands::browser`.
#[tauri::command]
pub fn browser_recipe_close(session_id: String) -> Result<String, String> {
    let mut map = instances().lock().map_err(|e| format!("Lock error: {}", e))?;

    if let Some(mut state) = map.remove(&session_id) {
        let msg = serde_json::json!({
            "id": 3,
            "action": "close",
            "params": {}
        });
        let _ = send_and_receive(&mut state, &msg.to_string(), 5000);
    }

    Ok("Recipe browser session closed".to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sanitize_profile_name_strips_path_traversal_and_separators() {
        assert_eq!(sanitize_profile_name("../../etc/passwd"), "etcpasswd");
        assert_eq!(sanitize_profile_name("my-brand_2"), "my-brand_2");
        assert_eq!(sanitize_profile_name(""), "default");
        assert_eq!(sanitize_profile_name("////"), "default");
        // Capped at 64 chars regardless of input length.
        let long = "a".repeat(200);
        assert_eq!(sanitize_profile_name(&long).len(), 64);
    }

    #[test]
    fn profile_dir_nests_sanitized_name_under_profiles_root() {
        let dir = profile_dir("Acme Brand!!");
        let name = dir.file_name().and_then(|n| n.to_str()).unwrap_or("");
        assert_eq!(name, "AcmeBrand");
        assert!(dir.parent().unwrap().ends_with("browser-profiles"));
    }

    #[test]
    fn resolve_step_secret_is_noop_without_secret_env_var() {
        let mut step = serde_json::json!({ "kind": "fill", "selector": "#x", "value": "literal" });
        resolve_step_secret(&mut step).expect("noop must not error");
        assert_eq!(step["value"], "literal");
    }

    #[test]
    fn resolve_step_secret_fails_fast_on_missing_env_var_without_leaking_a_value() {
        let mut step = serde_json::json!({
            "kind": "fill",
            "selector": "#password",
            "secretEnvVar": "LAZY_TEST_NONEXISTENT_SECRET_VAR_XYZ",
        });
        let err = resolve_step_secret(&mut step).expect_err("must fail when env var unset");
        assert!(err.contains("LAZY_TEST_NONEXISTENT_SECRET_VAR_XYZ"));
        assert!(!step.as_object().unwrap().contains_key("value"));
    }

    #[test]
    fn resolve_step_secret_substitutes_value_and_removes_the_env_var_reference() {
        std::env::set_var("LAZY_TEST_SECRET_VAR_ABC", "s3cr3t");
        let mut step = serde_json::json!({
            "kind": "fill",
            "selector": "#password",
            "secretEnvVar": "LAZY_TEST_SECRET_VAR_ABC",
        });
        resolve_step_secret(&mut step).expect("must resolve");
        assert_eq!(step["value"], "s3cr3t");
        assert!(!step.as_object().unwrap().contains_key("secretEnvVar"));
        std::env::remove_var("LAZY_TEST_SECRET_VAR_ABC");
    }
}
